### PR TITLE
Teswiz Upgrade | Appium 2 - iOS implementation 

### PR DESCRIPTION
**Pull request by mayankCloud**
Added Readme file for : iOS - Appium 2 Setup guide

### CODE REVIEW

File: docs/iOSAppium2SetupGuide-README.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+**Appium Dashboard plugin*** is a tool that allows to view and interact with iOS automation tests in real time. This can be helpful for debugging and troubleshooting tests.
+
+
+
+**Appium relaxed-caps plugin** is required because it allows to use capabilities that are not prefixed with the vendor's name. This can be helpful for compatibility reasons, as some vendors may not support the vendor prefix for certain capabilities.
+
+
+
+4. To install the Appium Device Farm and Appium Dashboard plugins, run the following commands.
```
Comment (anandbagmar) [inline docs/iOSAppium2SetupGuide-README.md:29]: the installation comments should be part of the description above

File: docs/iOSAppium2SetupGuide-README.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+
+
+## Start Appium
+
+
+
+10. To start Appium, run the following command:
+
+* `appium --relaxed-security --allow-cors --use-plugins="appium-dashboard,relaxed-caps,device-farm" -pa /wd/hub --plugin-device-farm-platform=both`
```
Comment (anandbagmar) [inline docs/iOSAppium2SetupGuide-README.md:74]: I am guessing this is to start appium manually. Let's make that explicit. Also highlight that teswiz will take care of this automatically

File: docs/iOSAppium2SetupGuide-README.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
+# Appium 2 – iOS Setup
+
+<span style="font-size: larger;">This guide will walk you through the steps to set up Appium 2 for iOS.</span> 
+
+1. Remove previous Appium 1.x installation if any.
+## Install Appium 2
+
+2. Install Appium 2 using the following command:
+- ```npm i -g appium@next```
```
[the remark this answers is not in the document]
Comment (mayankCloud) [reply docs/iOSAppium2SetupGuide-README.md:9]: Yes, I have checked the installation command is still the same.

File: docs/iOSAppium2SetupGuide-README.md (hunk 2 above)
Comment (mayankCloud) [reply docs/iOSAppium2SetupGuide-README.md:66]: Remark added.

File: build.gradle [hunk 4]
```diff
@@ -4,7 +4,7 @@ buildscript {
     ext {
         gradleVersion = '7.3.3'
         assertJVersion = '3.24.2'
-        atdVersion = '5b23b2273f'
+        atdVersion = '13.1.0'
```
Comment (anandbagmar) [inline build.gradle:7]: This is the correct ATD version - 5b23b2273f

Comment (mayankCloud) [reply build.gradle]: Updated to atdVersion to 5b23b2273f

File: caps/ajio_local_ios_capabilities.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{
+  "iOS": {
+    "platformName": "iOS",
+    "app": "src/test/resources/ajio.zip",
```
Comment (anandbagmar) [inline caps/ajio_local_ios_capabilities.json:4]: you can point to https://github.com/anandbagmar/sampleAppsForNativeMobileAutomation/ - if the app is no there, raise a PR to add it there

File: caps/ajio_local_capabilities.json [hunk 6]
```diff
@@ -41,5 +28,30 @@
         }
       ]
     }
-  ]
+  ],
+  "iOS": {
+    "platformName": "iOS",
+    "app": "src/test/resources/ajio.zip",
```
Comment (anandbagmar) [inline caps/ajio_local_capabilities.json:34]: you can point to https://github.com/anandbagmar/sampleAppsForNativeMobileAutomation/ - if the app is no there, raise a PR to add it there

File: configs/ajio_local_config.properties [hunk 7]
```diff
@@ -20,4 +20,5 @@ TARGET_ENVIRONMENT=prod
 LAUNCH_NAME_SUFFIX= on 'prod' Environment
 TEST_DATA_FILE=./src/test/resources/testData.json
 CLEANUP_DEVICE_BEFORE_STARTING_EXECUTION=false
-BROWSER_CONFIG_FILE=./configs/browser_config.json
\ No newline at end of file
+BROWSER_CONFIG_FILE=./configs/browser_config.json
+APPIUM_JS_PATH=/Users/mayank17.gupta/.appium/node_modules/appium/lib/main.js
```
Comment (anandbagmar) [inline configs/ajio_local_config.properties:24]: why is this required? Also, it cannot be a hardcoded path

File: docs/iOSAppium2SetupGuide-README.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+
+
+## Install Drivers
+
+
+
+5. Cross Platform installation, [click here](https://github.com/danielpaulus/go-ios), this is an operating system independent implementation of iOS device features. You can run UI tests, launch or kill apps, install apps etc. with it.
+   MacOS users will need to install this driver to unzip files. To install the driver, run the following command
+* `npm install -g go-ios`
```
Comment (anandbagmar) [inline docs/iOSAppium2SetupGuide-README.md:41]: remove '-g`

File: src/main/java/com/znsio/teswiz/runner/Setup.java [hunk 9]
```diff
@@ -133,7 +133,7 @@ static String getCurlProxyCommand() {
         return curlProxyCommand;
     }
 
-    static List<String> getExecutionArguments() {
+    static List<String> getExecutionArguments() throws IOException {
```
Comment (anandbagmar) [inline src/main/java/com/znsio/teswiz/runner/Setup.java:136]: why are we throwing exceptions? Throw a custom runtime exception

File: src/main/java/com/znsio/teswiz/runner/Runner.java [hunk 10]
```diff
@@ -37,7 +38,7 @@ public Runner() {
         throw new InvalidTestDataException("Required args not provided to Runner");
     }
 
-    public Runner(String configFilePath, String stepDefDirName, String featuresDirName) {
+    public Runner(String configFilePath, String stepDefDirName, String featuresDirName) throws IOException {
```
Comment (anandbagmar) [inline src/main/java/com/znsio/teswiz/runner/Runner.java:41]: why are we throwing exceptions? Throw a custom runtime exception

File: src/main/java/com/znsio/teswiz/runner/LocalDevicesSetup.java [hunk 11]
```diff
@@ -71,4 +76,49 @@ private static String getAdbCommandOutputFromLocalDevice(JadbDevice device, Stri
         LOGGER.info("\tOutput: " + adbCommandOutput);
         return adbCommandOutput;
     }
+
+    private static List<String> getBootedIOSSimulators() throws IOException {
```
Comment (anandbagmar) [inline src/main/java/com/znsio/teswiz/runner/LocalDevicesSetup.java:80]: why are we throwing exceptions? Throw a custom runtime exception

File: src/main/java/com/znsio/teswiz/runner/DeviceSetup.java [hunk 12]
```diff
@@ -261,4 +271,26 @@ static void cleanupCloudExecution() {
                         String.format("Provided cloudName: '%s' is not supported", cloudName));
         }
     }
+
+    static ArrayList<String> setupIOSExecution() throws IOException {
```
Comment (anandbagmar) [inline src/main/java/com/znsio/teswiz/runner/DeviceSetup.java:275]: why are we throwing exceptions? Throw a custom runtime exception

File: src/main/java/com/znsio/teswiz/runner/AppiumDriverManager.java [hunk 13]
```diff
@@ -422,4 +424,102 @@ private static Driver createWindowsDriver(String userPersona, Platform forPlatfo
...
+
+    static void closeIOSAppOnDevice(String userPersona,
+                                    @NotNull
+                                    Driver driver) {
+        String appPackageName = Runner.getAppPackageName();
+        String logMessage;
+        --numberOfAppiumDriversUsed;
+        LOGGER.info(String.format("numberOfAppiumDriversUsed: %d", numberOfAppiumDriversUsed));
+        AppiumDriver appiumDriver = (AppiumDriver) driver.getInnerDriver();
+        if (null == appiumDriver) {
```
Comment (anandbagmar) [inline src/main/java/com/znsio/teswiz/runner/AppiumDriverManager.java:475]: break if-else into smaller methods

File: src/test/java/com/znsio/teswiz/screen/ios/ajio/CartScreenIOS.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    @Override
+    public String getActualProductName() {
+        LOGGER.info("getCartProductName");
+        waitFor(5);
+        if (driver.isElementPresent(byWishlistPopUp)) {
+            driver.waitTillElementIsPresent(byWishlistPopUp).click();
+        }
+        visually.checkWindow(SCREEN_NAME, "Product in the cart");
```
Comment (anandbagmar) [inline src/test/java/com/znsio/teswiz/screen/ios/ajio/CartScreenIOS.java:34]: add visual validations

File: src/test/java/com/znsio/teswiz/screen/ios/ajio/HomeScreenIOS.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    private final Visual visually;
+
+    public HomeScreenIOS(Driver driver, Visual visually) {
+        this.driver = driver;
+        this.visually = visually;
+    }
+
+    @Override
+    public SearchScreen searchByImage() {
+        driver.waitTillElementIsPresent(byImageId).click();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/teswiz/screen/ios/ajio/HomeScreenIOS.java:29]: add visual validation

File: src/test/java/com/znsio/teswiz/screen/ios/ajio/ProductScreenIOS.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public CartScreen addProductToCart() {
+        LOGGER.info("addProductToCart");
+        driver.waitTillElementIsPresent(byAddToCartButtonId).click();
+        visually.checkWindow(SCREEN_NAME, "Add to cart");
+        driver.waitTillElementIsPresent(bySelectSizeMediumXpath).click();
+        driver.waitTillElementIsPresent(byAddToBagId).click();
+        driver.waitTillElementIsPresent(byViewBagButtonId).click();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/teswiz/screen/ios/ajio/ProductScreenIOS.java:37]: add visual validation

File: caps/ajio_local_capabilities.json [hunk 17]
```diff
@@ -41,5 +28,30 @@
...
+    "simpleIsVisibleCheck": true,
+    "maxTypingFrequency": 10,
+    "automationName": "XCUITest",
+    "usePrebuiltWDA": true,
+    "appium:iPhoneOnly": true,
+    "noReset": true
+  },
+  "serverConfig" : {
+    "server": {
+      "use-plugins": ["device-farm", "appium-dashboard","relaxed-caps-plugin"],
```
[the remark this answers is not in the document]
Comment (mayankCloud) [reply caps/ajio_local_capabilities.json:48]: It is optional, I only added it as it increases compatibility for support of legacy apps and also if vendors does not provides all the required capabilities, with this plugin we can still connect both iOS and Android apps.

File: configs/ajio_local_config.properties (hunk 7 above)
Comment (mayankCloud) [reply configs/ajio_local_config.properties:24]: This APPIUM_JS_PATH config is required as part of ATD, without this test case will not run. Shall I remove my username and enter "Your system UserName" and then push the change. Please suggest.

File: configs/ajio_local_ios_config.properties [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+PLATFORM=iOS
+PROXY_KEY=HTTP_PROXY
+REPORT_PORTAL_FILE=src/test/resources/reportportal.properties
+RUN_IN_CI=false
+TARGET_ENVIRONMENT=prod
+LAUNCH_NAME_SUFFIX= on 'prod' Environment
+TEST_DATA_FILE=./src/test/resources/testData.json
+CLEANUP_DEVICE_BEFORE_STARTING_EXECUTION=false
+BROWSER_CONFIG_FILE=./configs/browser_config.json
+APPIUM_JS_PATH=/Users/mayank17.gupta/.appium/node_modules/appium/lib/main.js
```
[the remark this answers is not in the document]
Comment (mayankCloud) [reply configs/ajio_local_ios_config.properties:20]: This APPIUM_JS_PATH config is required as part of ATD, without this test case will not run. Shall I remove my username and enter "Your system UserName" and then push the change. Please suggest.

This is same as "ajio_local_config.properties" I have created a separate copy to avoid confusion the local properties is capable of running both Android and iOS tests. However in order to avoid confusion for iOS setup i created this copy.

File: caps/ajio_local_capabilities.json (hunk 6 above)
Comment (mayankCloud) [reply caps/ajio_local_capabilities.json:34]: PR raised

File: caps/ajio_local_ios_capabilities.json (hunk 5 above)
Comment (mayankCloud) [reply caps/ajio_local_ios_capabilities.json:4]: PR raised

File: src/main/java/com/znsio/teswiz/runner/DeviceSetup.java (hunk 12 above)
Comment (mayankCloud) [reply src/main/java/com/znsio/teswiz/runner/DeviceSetup.java]: Not required here, handled in the calling method verifyAppExistsAtMentionedPath();

File: src/main/java/com/znsio/teswiz/runner/AppiumDriverManager.java (hunk 13 above)
Comment (mayankCloud) [reply src/main/java/com/znsio/teswiz/runner/AppiumDriverManager.java:474]: Optimised.

File: docs/iOSAppium2SetupGuide-README.md [hunk 19]
```diff
@@ -0,0 +1,116 @@
+# Appium 2 – iOS Setup
+
+<span style="font-size: larger;">This guide will walk you through the steps to set up Appium 2 for iOS.</span> 
+
+1. Remove previous Appium 1.x installation if any.
+## Install Appium 2
+
+2. Install Appium 2 using the following command:
+- ```npm i -g appium@next```
```
[the remark this answers is not in the document]
Comment (mayankCloud) [reply docs/iOSAppium2SetupGuide-README.md]: Yes, tested on teswiz and ATD

File: src/main/java/com/znsio/teswiz/runner/Setup.java (hunk 9 above)
Comment (mayankCloud) [reply src/main/java/com/znsio/teswiz/runner/Setup.java]: Removed it got added because of method chaining, have fixed the method from where it was getting generated.

File: src/main/java/com/znsio/teswiz/runner/Setup.java [hunk 20]
```diff
@@ -231,10 +231,11 @@ private static Map<String, Map> loadTestDataForEnvironment(String environment) {
                                                                 environment, testDataFile);
     }
 
-    private static void setupExecutionEnvironment() {
+    private static void setupExecutionEnvironment() throws IOException {
```
[the remark this answers is not in the document]
Comment (mayankCloud) [reply src/main/java/com/znsio/teswiz/runner/Setup.java]: Removed it got added because of method chaining, have fixed the method from where it was getting generated.

File: src/main/java/com/znsio/teswiz/runner/Runner.java [hunk 21]
```diff
@@ -125,7 +126,7 @@ public static Map<String, Object> getTestDataAsMap(String key) {
         }
     }
 
-    public static void main(String[] args) {
+    public static void main(String[] args) throws IOException {
```
[the remark this answers is not in the document]
Comment (mayankCloud) [reply src/main/java/com/znsio/teswiz/runner/Runner.java]: Removed it got added because of method chaining, have fixed the method from where it was getting generated.

File: src/main/java/com/znsio/teswiz/runner/Runner.java (hunk 10 above)
Comment (mayankCloud) [reply src/main/java/com/znsio/teswiz/runner/Runner.java]: Removed it got added because of method chaining, have fixed the method from where it was getting generated.

File: src/main/java/com/znsio/teswiz/runner/LocalDevicesSetup.java [hunk 22]
```diff
@@ -71,4 +76,49 @@ private static String getAdbCommandOutputFromLocalDevice(JadbDevice device, Stri
...
+                System.err.println("Failed to fetch iOS Simulators. Exit code: " + exitCode);
+            }
+        } catch (IOException | InterruptedException e) {
+            throw new EnvironmentSetupException("No simulators detected to run the tests");
+        }
+
+        return bootedSimulators;
+    }
+
+    static void setupLocalIOSExecution() throws IOException {
```
[the remark this answers is not in the document]
Comment (mayankCloud) [reply src/main/java/com/znsio/teswiz/runner/LocalDevicesSetup.java]: Removed it got added because of method chaining, have fixed the method from where it was getting generated.

File: src/main/java/com/znsio/teswiz/runner/LocalDevicesSetup.java (hunk 11 above)
Comment (mayankCloud) [reply src/main/java/com/znsio/teswiz/runner/LocalDevicesSetup.java]: Removed it got added because of method chaining, have fixed the method from where it was getting generated.

File: docs/iOSAppium2SetupGuide-README.md (hunk 8 above)
Comment (mayankCloud) [reply docs/iOSAppium2SetupGuide-README.md]: updated, '-g' removed.

File: src/test/java/com/znsio/teswiz/screen/ios/ajio/CartScreenIOS.java (hunk 14 above)
Comment (mayankCloud) [reply src/test/java/com/znsio/teswiz/screen/ios/ajio/CartScreenIOS.java]: Visual validation step added.

File: src/test/java/com/znsio/teswiz/screen/ios/ajio/HomeScreenIOS.java (hunk 15 above)
Comment (mayankCloud) [reply src/test/java/com/znsio/teswiz/screen/ios/ajio/HomeScreenIOS.java]: Visual validation step added.

File: src/test/java/com/znsio/teswiz/screen/ios/ajio/ProductScreenIOS.java (hunk 16 above)
Comment (mayankCloud) [reply src/test/java/com/znsio/teswiz/screen/ios/ajio/ProductScreenIOS.java:38]: Visual validation step added.

File: caps/ajio_local_capabilities.json (hunk 6 above)
Comment (mayankCloud) [reply caps/ajio_local_capabilities.json:34]: As discussed we will not be pushing it in repo as the app is internal to AJIO.

File: caps/ajio_local_ios_capabilities.json (hunk 5 above)
Comment (mayankCloud) [reply caps/ajio_local_ios_capabilities.json:4]: As discussed we will not be pushing it in repo as the app is internal to AJIO.
